libraries/chain/block_log.cpp

```diff
@@ -477,22 +483,256 @@ namespace eosio { namespace chain {
+               }
+            }
+            this->collection.erase(this->collection.begin(), this->collection.begin() + items_to_erase);
+            this->active_index = this->active_index == npos  || this->active_index < items_to_erase ? npos : this->active_index - items_to_erase;
+         }
+         if (max_retained_files > 0)
+            this->collection.emplace(start_block_num, mapped_type{end_block_num, filename_base});
+      }
+   };
+
+   namespace detail {
+
+      /**
+       * The implementation detail for the read/write access to the block log/index
+       *
+       * @note All the non-static member functions require to fullfill the class invariant after execution unless
+       *exceptions are thrown.
+       * @invariant block_file.is_open() && index_file.is_open()
+       **/
+      class block_log_impl {
+       public:
+         signed_block_ptr          head;
+         block_log_catalog         catalog;
+         fc::datastream<fc::cfile> block_file;
+         fc::datastream<fc::cfile> index_file;
+         bool                      genesis_written_to_block_log = false;
+         block_log_preamble        preamble;
+         size_t                    stride = std::numeric_limits<size_t>::max();
+         static uint32_t           default_version;
+
+         block_log_impl(const fc::path& data_dir, fc::path archive_dir, uint64_t stride, uint16_t max_retained_files);
+
+         static void ensure_file_exists(fc::cfile& f) {
+            if (fc::exists(f.get_file_path()))
+               return;
+            f.open(fc::cfile::create_or_update_rw_mode);
+            f.close();
+         }
+
+         uint64_t get_block_pos(uint32_t block_num);
+
+         void reset(uint32_t first_block_num, std::variant<genesis_state, chain_id_type>&& chain_context);
+
+         void flush();
 
-   block_log::~block_log(){}
+         uint64_t append(const signed_block_ptr& b, packed_transaction::cf_compression_type segment_compression);
+
+         uint64_t write_log_entry(const signed_block& b, packed_transaction::cf_compression_type segment_compression);
+
+         void split_log();
+         bool recover_from_incomplete_block_head(block_log_data& log_data, block_log_index& index);
+
+         block_id_type                 read_block_id_by_num(uint32_t block_num);
+         std::unique_ptr<signed_block> read_block_by_num(uint32_t block_num);
+         void                          read_head();
+      };
+      uint32_t block_log_impl::default_version = block_log::max_supported_version;
+   } // namespace detail
+
+   block_log::block_log(const fc::path& data_dir, fc::path archive_dir, uint64_t stride,
+                        uint16_t max_retained_files)
+       : my(new detail::block_log_impl(data_dir, archive_dir, stride, max_retained_files)) {}
+
+   block_log::~block_log() {}
+
+   bool detail::block_log_impl::recover_from_incomplete_block_head(block_log_data& log_data, block_log_index& index) {
```

Comment:
@b1bart we need to discus this, it is a change to how nodeos functions with an invalid block log.